### PR TITLE
Toggle between Rectangles and Markers on map

### PR DESCRIPTION
**Pull request by sean-prz**
_closes #94_
### Toggle between simple and advance mode on map screen

- The user can now choose if he wants to see markers or more detailed rectangles. 
- The pin on the map has been changed to not be a default red marker anymore

### How

- introduce a state that holds the mode of the map.
- Mode of the map are represented by an enum added in this PR
- added a toggle switch that changes the map mode state.
- when the map is too unzoomed the state is also automatically changed and the switch hidden on the UI.
- the state is observed and  either call `drawMarkers` or `drawRectangles`

A good portion of the lines changed in this PR are just reordering and grouping lines of code in the map screen for more readability.
-- -
### Demo. 
https://github.com/user-attachments/assets/0cfdb076-c1e7-456a-86a8-4ade0861c7da
-- -
### Code coverage 
image

### CODE REVIEW

**Comment by AlexandreBeeDeeBop** [comment]
I havent read the code yet so maybe this is not your problem but here's a few things I noticed while on the emulator on your branch :

- The little button to toggle Advanced Mode could be improved by
1. having a different color if it's not toggled (maybe white) than when it's toggled (maybe blue like it is right now for both modes).
2. having the text above it be Normal mode when it's not toggled and Advanced mode when it's toggled (like it is right now for both modes).

- When you activate advanced mode -> dezoom so that "when the map is too unzoomed the state is also automatically changed and the switch hidden on the UI." as you say -> and zoom back, you are not in advanced mode anymore, which can be annoying.


That's it for now, the feature is really cool and now im going to see if I can find issues in the code itself.

**Comment by sean-prz** [comment]
> I havent read the code yet so maybe this is not your problem but here's a few things I noticed while on the emulator on your branch :
> 
> * The little button to toggle Advanced Mode could be improved by
> 
> 1. having a different color if it's not toggled (maybe white) than when it's toggled (maybe blue like it is right now for both modes).
> 2. having the text above it be Normal mode when it's not toggled and Advanced mode when it's toggled (like it is right now for both modes).
> 
> * When you activate advanced mode -> dezoom so that "when the map is too unzoomed the state is also automatically changed and the switch hidden on the UI." as you say -> and zoom back :
>   you are not in advanced mode anymore, which can be annoying.
> 
> That's it for now, the feature is really cool and now im going to see if I can find issues in the code itself.

I totally with prop 1, but I am waiting for the theme to be defined, because I don't want to put any hardcoded value so for now it's just PrimaryColor which is the only one that is not ugly.

I don't really get point 3 can you rephrase ?

**Comment by AlexandreBeeDeeBop** [comment]
Ok fair enough maybe we dont need to have the text be changed, but the color yes we agree on that.
Ill explain how you can test my points 3 :
Zoom to the max, activate advanced mode, dezoom to the max, zoom to the max. You are now in normal modem not in advanced mode anymore. I give you the extreme one so you see the problem, the thing is that even if you dont exaggerate by max zooming then max dezooming etc, there's still this one moment when a single click on the - followed by a single click on the + can make you go from advanced mode to normal mode, which might by annoying for users

**Comment by sean-prz** [comment]
OK I understand now, I can try to add another state to remember the user selectionb

**Comment by AlexandreBeeDeeBop** [comment]
Ok good, it's because of this feature you added isnt it ? :

"when the map is too unzoomed the state is also automatically changed and the switch hidden on the UI."

**Comment by AlexandreBeeDeeBop** [comment]
ok, can you add a test that zooms, activates advanced mode, unzooms, zooms, asserts advanced mode is still activated ?

**Comment by sean-prz** [comment]
Complicated, because every state is internal. I can extend the mapscreen test to assert that the toggle switch is there

**Comment by AlexandreBeeDeeBop** [comment]
nevermind dont bother we can test it ourselves with the emulator. I was thinking about the test coverage.

File: app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt [hunk 1]
```diff
@@ -249,10 +310,25 @@ fun MapScreen(
...
+
+                  // store the zoom level
+                  // This must stay at the end of the listener.
+                  zoomState.value = mapView.mapboxMap.cameraState.zoom
                 }
+            // =======================  CAMERA  LISTENER  =======================
 
             onDispose {
-              pointAnnotationManager?.deleteAll()
+              markerAnnotationManager?.deleteAll()
```
Comment (Armetiis) [inline app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt:331]: here shouldn't we delete all annotations of all kind? (rectangle and pLabel)

File: app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt [hunk 2]
```diff
@@ -261,6 +337,38 @@ fun MapScreen(
 
     // ======================= OVERLAY =======================
     Box(modifier = Modifier.padding(padding).fillMaxSize()) {
+      // A switch to change the display mode
+      Column(
+          modifier =
+              Modifier.align(Alignment.TopStart).padding(start = 8.dp, top = 32.dp).alpha(alpha),
+          horizontalAlignment = Alignment.CenterHorizontally) {
+            Text(
+                "Advanced Mode",
```
Comment (Armetiis) [inline app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt:346]: Please use string resources

File: app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt [hunk 3]
```diff
@@ -319,26 +418,40 @@ fun MapScreen(
...
       val polygonAnnotationOptions =
           PolygonAnnotationOptions()
               .withGeometry(polygon)
-              .withFillColor("#22799B")
+              .withFillColor("#1A4988")
               .withFillOpacity(0.7)
-      polygonAnnotationManager?.create(polygonAnnotationOptions)
+      annotations.add(polygonAnnotationOptions)
+      val area = TurfMeasurement.area(polygon)
+      Log.d("MapScreen", "Area: $area")
```
Comment (Armetiis) [inline app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt:440]: I think you forgot to remove the log here

File: app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt [hunk 4]
```diff
@@ -71,12 +80,28 @@ import com.mapbox.maps.viewannotation.annotatedLayerFeature
...
 import com.mapbox.maps.viewannotation.geometry
 import com.mapbox.maps.viewannotation.viewAnnotationOptions
+import com.mapbox.turf.TurfMeasurement
 
 const val defaultZoom = 16.0
 const val maxZoom = 18.0
 const val minZoom = 8.0
 const val thresholdDisplayZoom = 13.0
 const val LAYER_ID = "0128"
+const val LABEL_THRESHOLD = 15.5
```
Comment (Armetiis) [inline app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt:90]: In my opinion this name is not very clear. Maybe something like "Advanced mode zoom treshold" would be clearer?

**Review by Armetiis — APPROVED** [review]
Really great job you did here. Everything looks great. 
Apart from the comments I left you in the code I have 2 suggestions

1. As Alexandre said, having the toggle switch gray when disabled would really improve it. Also, I think it would look better if the text and the toggle were one next to the other (in a row rather than column). You can remove "Mode" in "Advanced Mode" so the row isn't too long

2. I don't know how much work this might imply, so if you consider it too hard/long, this can be the subject of a future PR. Otherwise, you think you can make changes so that clicking on a parking rectangle does the same as clicking on a ping? (Display the preview card)

None of this is blocking, so you have my approval. However, these little changes could make a difference. Up to you.
Also try to update your coverage after rebasing

PS: Thank you very much for taking the time to refactor the file

File: app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt (hunk 1 above)
Comment (sean-prz) [reply app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt:331]: Fixed

File: app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt (hunk 2 above)
Comment (sean-prz) [reply app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt:346]: Fixed

File: app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt (hunk 4 above)
Comment (sean-prz) [reply app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt:90]: Fixed

File: app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt (hunk 3 above)
Comment (sean-prz) [reply app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt:440]: Fixed

**Comment by sean-prz** [comment]
Thanks for all the very pertinent comments, I fixed them all.

**Comment by Armetiis** [comment]
Prop 1 is perfect!

**Review by AlexandreBeeDeeBop — APPROVED** [review]
Once you're done adressing all the changes requested you're good to merge.
